File: src/core/implementation/service/sync_version_manager.ts
import { SyncalbeMetadataRepository } from "../data/sync_metadata_repository";
import { ISyncVersionManager } from "../../abstraction/service/ISync_version_manager";
import { SyncMetadata } from "../metadata/syncable_metadata";
import { ISyncalbeDataSource } from "../../abstraction/data/ISyncable_data_source";
import { isNullOrUndefined } from "../utils/utils";

export class SyncVersionManager implements ISyncVersionManager {

    private syncMetadataRepository: SyncalbeMetadataRepository<SyncMetadata>;
    constructor(private syncMetadataDataSource : ISyncalbeDataSource<SyncMetadata>){
        this.syncMetadataRepository = new SyncalbeMetadataRepository(this.syncMetadataDataSource)
    }


    public async getLastSyncVersion(entityType: string, syncZone: string): Promise<number> {
        let syncMetadataList = await this.syncMetadataRepository.query({type :entityType, syncZone: syncZone});
        if(syncMetadataList && syncMetadataList.length != 0){
            return syncMetadataList[0].version;
        }
        else{
            let syncMetadata = await this.initObjectMetadata(entityType,syncZone)
            return syncMetadata.version
        }
    }

    public async incrementSyncVersion(entityType: string, syncZone: string): Promise<number> {
        let syncMetadataList = await this.syncMetadataRepository.query({type :entityType,syncZone: syncZone});
        let syncMetadata : SyncMetadata;
        if(syncMetadataList && syncMetadataList.length != 0){
            syncMetadata = syncMetadataList[0];
        }
        else{
             syncMetadata = await this.initObjectMetadata(entityType,syncZone)
        }
        syncMetadata.version ++;
        syncMetadata = await this.syncMetadataRepository.update({_id: syncMetadata._id},syncMetadata);
        return syncMetadata.version;
    }

    private async initObjectMetadata(entityType: string, syncZone: string): Promise<SyncMetadata>{
        let syncMetadataList = await this.syncMetadataRepository.query({type :entityType, syncZone: syncZone});
        if(isNullOrUndefined(syncMetadataList) || syncMetadataList.length == 0){
            let syncMetadata = new SyncMetadata(entityType, 0, syncZone);
            syncMetadata.timestamp = new Date().getTime();
            return await this.syncMetadataRepository.add(syncMetadata);        
        }
    }
} 

File: fast_sync_dotnet_service/fast-sync-core/implementation/service/SyncVersionManager.cs
using fast_sync_core.abstraction.data;
using fast_sync_core.implementation.data;
using fast_sync_core.implementation.metadata;

namespace fast_sync_core.implementation
{
    public class SyncVersionManager : ISyncVersionManager

    {
        public ISyncableDataSource<SyncMetadata> metadataDataSource;

        public SyncVersionManager(ISyncableDataSource<SyncMetadata> syncMetadataDataSource)
        {
            metadataDataSource = syncMetadataDataSource;
        }

        public async Task<int> GetLastSyncVersion(string entityType, string syncZone)
        {
            var syncMetadataList = await metadataDataSource.Query((metadata)=> metadata.Type == entityType && metadata.SyncZone == syncZone);
            if (syncMetadataList != null && syncMetadataList.Count != 0)
            {
                return syncMetadataList[0].Version;
            }
            else
            {
                var syncMetadata = await InitObjectMetadata(entityType, syncZone);
                return syncMetadata.Version;
            }
        }

        public async Task<int> IncrementSyncVersion(string entityType, string syncZone)
        {
            List<SyncMetadata> syncMetadataList = await metadataDataSource.Query((metadata) => metadata.Type == entityType && metadata.SyncZone == syncZone);
            SyncMetadata syncMetadata;
            if (syncMetadataList != null && syncMetadataList.Count != 0)
            {
                syncMetadata = syncMetadataList[0];
            }
            else
            {
                syncMetadata = await InitObjectMetadata(entityType, syncZone);
            }
            syncMetadata.Version++;
            syncMetadata = await metadataDataSource.Update(syncMetadata.Id, syncMetadata);
            return syncMetadata.Version;
        }

        private async Task<SyncMetadata> InitObjectMetadata(string entityType, string syncZone)
        {
            var syncMetadataList = await metadataDataSource.Query((metadata) => metadata.Type == entityType && metadata.SyncZone == syncZone);
            if (syncMetadataList == null || syncMetadataList.Count == 0)
            {
                var syncMetadata = new SyncMetadata();
                syncMetadata.SyncZone = syncZone;
                syncMetadata.Version = 0;
                syncMetadata.Type = entityType;
                syncMetadata.Timestamp = DateTime.Now.Ticks;
                return await metadataDataSource.Add(syncMetadata);
            }
            return syncMetadataList.First();
        }
    }

}
